pkg/api/types.go

```diff
@@ -2070,10 +2070,41 @@ const (
 	StrategicMergePatchType PatchType = "application/strategic-merge-patch+json"
 )
 
+// Type and constants for component type.
+type ComponentType string
+
+// These are the valid conditions for the component type.
+const (
+	ComponentAPIServer         ComponentType = "apiserver"
+	ComponentControllerManager ComponentType = "controller-manager"
+	ComponentScheduler         ComponentType = "scheduler" // TODO: move mesos-specific component type & validation to contrib
+)
+
+// Component describes metadata about a core component.
+type Component struct {
+	TypeMeta   `json:",inline"`
+	ObjectMeta `json:"metadata,omitempty"`
+
+	// Type of the component
+	Type ComponentType `json:"type"`
+	// URL to reach the component (scheme, host, port, path)
+	URL string `json:"location"`
+
+	// The time at which the component was last updated.
+	LastTimestamp util.Time `json:"lastTimestamp,omitempty"`
```

Comment:
I had originally thought that Status was unnecessary, because a heartbeating component is always healthy (otherwise it wouldn't be heartbeating), but after more thought I think it's reasonable for a component to continue heartbeating when unhealthy when one of its dependencies is unhealthy. For example, apiserver should remain alive, but unhealthy, when etcd is down. That way it can auto-recover when etcd recovers. So I think I'll add in Status.

This does mean, however, that the status could be a lie if the component died and stopped heartbeating. At some point the LastTimestamp needs to time out and trigger Status=Unhealthy. But I'm not sure where to put that logic.

I suppose I could create a timer goroutine on every update with the configured timeout. If another update is received, the timer is stopped. If no update is received, the timer triggers a status change. I'm not sure how this would work in an HA apiserver tho...